### PR TITLE
rework unit tests (mocking the fetch())

### PR DESCRIPTION
**Pull request by ftoromanoff**
This PR propose a rewrite of the unit tests:
- minor corrections in tests
- mocking of the 'fetch()' function  to have unit tests to run independent from network
    - to do that the files that should have been fetch have been added localy.
    - we use sinon to mock itowns.Fetcher
- add a file test for testing Fetcher.js

Modifications:
- add 'sinon' library to do the mocking
- all tests
  -  add all test files in local folder test/data/unitTest
  - change global url to local ones
- add fetcher.js to test Fetcher.js
  - add library xml dom for xml fetcher

### CODE REVIEW

File: test/unit/3dtileslayerprocessbatchtable.js [hunk 1]
```diff
@@ -8,34 +9,58 @@ import View from 'Core/View';
 import GlobeView from 'Core/Prefab/GlobeView';
 import { HttpsProxyAgent } from 'https-proxy-agent';
 import Coordinates from 'Core/Geographic/Coordinates';
+import Fetcher from 'Provider/Fetcher';
+import sinon from 'sinon';
 import Renderer from './bootstrap';
 
+import tilesetBatchTable from '../data/3dTiles/tilesetBatchTable.json';
+
+const tileBatchTable = fs.readFileSync('./test/data/3dTiles/tile.b3dm');
```
Comment (Desplandis) [inline test/unit/3dtileslayerprocessbatchtable.js]: Use a relative path.

```suggestion
const tileBatchTable = fs.readFileSync('../data/3dTiles/tile.b3dm');
```

Comment (Desplandis) [inline test/unit/3dtileslayerprocessbatchtable.js]: I think we could reorganize the data folder, since those dataset comes from [Cesium](https://github.com/CesiumGS/cesium/tree/main/Apps/SampleData/Cesium3DTiles/Hierarchy/BatchTableHierarchy), we could move both `tilesetBatchTable.json` and `tile.b3dm` in a subfolder, maybe `batchTableHierarchy`?

File: test/unit/3dtileslayerstyle.js [hunk 2]
```diff
@@ -7,79 +7,104 @@ import PlanarView from 'Core/Prefab/PlanarView';
...
+        }
+    }
+    geometry.setAttribute('_BATCHID', new THREE.BufferAttribute(Int32Array.from(array), 1));
+
+    const result = new THREE.Mesh(geometry, material);
+    result.batchTable = new C3DTBatchTable();
+    result.tileId = tileId;
+
+    return result;
+}
```
Comment (Desplandis) [inline test/unit/3dtileslayerstyle.js:38]: If this is used some in some other places, we could move it in some stub file? We'll maybe do it in another PR.

File: test/unit/3dtileslayerstyle.js [hunk 3]
```diff
@@ -7,79 +7,104 @@ import PlanarView from 'Core/Prefab/PlanarView';
...
-        for (let index = 0; index < geometry.attributes.position.count; index++) {
-            array.push(currentBatchId);
-
-            // Change randomly batch id
-            if (Math.random() > 0.5) {
-                currentBatchId = Math.round(Math.random() * 50);
-            }
-        }
-        geometry.setAttribute('_BATCHID', new THREE.BufferAttribute(Int32Array.from(array), 1));
+            networkOptions: process.env.HTTPS_PROXY ? { agent: new HttpsProxyAgent(process.env.HTTPS_PROXY) } : {},
```
Comment (Desplandis) [inline test/unit/3dtileslayerstyle.js]: Not necessary anymore since our tests are now local.

```suggestion
```

File: test/unit/3dtileslayerstyle.js [hunk 4]
```diff
@@ -7,79 +7,104 @@ import PlanarView from 'Core/Prefab/PlanarView';
 import C3DTBatchTable from 'Core/3DTiles/C3DTBatchTable';
 import C3DTilesSource from 'Source/C3DTilesSource';
 import C3DTilesLayer from 'Layer/C3DTilesLayer';
+import sinon from 'sinon';
+import Fetcher from 'Provider/Fetcher';
 import Renderer from './bootstrap';
 
+import tileset from '../data/3dTiles/tileset.json';
```
Comment (Desplandis) [inline test/unit/3dtileslayerstyle.js]: This file is 40kb and could be removed. Seems we can instantiate a `C3DTilesSource` with an empty tileset in iTowns.

```suggestion
const tileset = {};
```

File: test/unit/3dtileslayerstyle.js [hunk 5]
```diff
@@ -7,79 +7,104 @@ import PlanarView from 'Core/Prefab/PlanarView';
...
-    // Define crs
-    proj4.defs('EPSG:3946', '+proj=lcc +lat_1=45.25 +lat_2=46.75 +lat_0=46 +lon_0=3 +x_0=1700000 +y_0=5200000 +ellps=GRS80 +towgs84=0,0,0,0,0,0,0 +units=m +no_defs');
+    let view;
+    let $3dTilesLayer;
+    let source;
+    let stubFetcherJson;
+
+    before(function () {
+        stubFetcherJson = sinon.stub(Fetcher, 'json')
+            .callsFake(() => Promise.resolve(JSON.parse(tileset)));
```
Comment (Desplandis) [inline test/unit/3dtileslayerstyle.js]: Given the top comment, we could change by:

```suggestion
            .callsFake(() => Promise.resolve(tileset));
```

File: test/unit/featuregeometrylayer.js [hunk 6]
```diff
@@ -3,71 +3,87 @@ import assert from 'assert';
...
 import { HttpsProxyAgent } from 'https-proxy-agent';
 import Extent from 'Core/Geographic/Extent';
 import Coordinates from 'Core/Geographic/Coordinates';
 import OBB from 'Renderer/OBB';
 import TileMesh from 'Core/TileMesh';
-import Style from 'Core/Style';
+import sinon from 'sinon';
 import Renderer from './bootstrap';
 
+import feature from '../data/filesource/feat_Polygone.geojson';
```
Comment (Desplandis) [inline test/unit/featuregeometrylayer.js:14]: Move to `../data/geojson`.

File: test/unit/geoidlayer.js [hunk 7]
```diff
@@ -2,43 +2,70 @@ import * as THREE from 'three';
...
 import Coordinates from 'Core/Geographic/Coordinates';
 import GlobeView from 'Core/Prefab/GlobeView';
 import { HttpsProxyAgent } from 'https-proxy-agent';
 import Extent from 'Core/Geographic/Extent';
 import OBB from 'Renderer/OBB';
 import TileMesh from 'Core/TileMesh';
+import sinon from 'sinon';
 import Renderer from './bootstrap';
 
+import createGtxBuffer from './gtxparser';
```
Comment (Desplandis) [inline test/unit/geoidlayer.js]: We should have a `utils` file for this sort of stub creation.

File: test/unit/gtxparser.js [hunk 8]
```diff
@@ -1,20 +1,33 @@
...
 import GTXParser from 'Parser/GTXParser';
-import Fetcher from 'Provider/Fetcher';
 
+const HEADER_SIZE = 40;
+const GTX_ROWS = 381;
+const GTX_COLS = 421;
 
-describe('GTXParser', function () {
-    let buffer;
+export default function createGtxBuffer(elevation = 1) {
```
Comment (Desplandis) [inline test/unit/gtxparser.js]: Maybe we should aggregate all data stubs in a unique file?

File: test/unit/potree.js [hunk 9]
```diff
@@ -8,27 +9,52 @@ import Coordinates from 'Core/Geographic/Coordinates';
...
+import sinon from 'sinon';
+import Fetcher from 'Provider/Fetcher';
 import Renderer from './bootstrap';
 
+import fileJson from '../data/potree/eglise_saint_blaise_arles.json';
+
+const urlPotreeFile = 'https://raw.githubusercontent.com/gmaillet/dataset/master';
+const fileName = 'eglise_saint_blaise_arles';
+
+const potreeRRhrc = fs.readFileSync('./test/data/potree/eglise_saint_blaise_arles/r/r.hrc');
```
Comment (Desplandis) [inline test/unit/potree.js]: Those files have a total size of 300kB. Depending on whether it is still possible to generate potree 1.0 files, I'll generate a new dataset.

File: test/unit/potree.js [hunk 10]
```diff
@@ -8,27 +9,52 @@ import Coordinates from 'Core/Geographic/Coordinates';
...
+            .callsFake(() => Promise.resolve(JSON.parse(fileJson)));
+        stubFetcherArrayBuf = sinon.stub(Fetcher, 'arrayBuffer')
+            .callsFake(url => Promise.resolve(resources.get(url)));
         renderer = new Renderer();
         viewer = new GlobeView(renderer.domElement, placement, { renderer });
 
+        const source = new PotreeSource({
+            file: 'eglise_saint_blaise_arles.json',
+            url: urlPotreeFile,
+            networkOptions: process.env.HTTPS_PROXY ? { agent: new HttpsProxyAgent(process.env.HTTPS_PROXY) } : {},
```
Comment (Desplandis) [inline test/unit/potree.js]:
```suggestion
```

File: test/unit/potreelayerparsing.js [hunk 11]
```diff
@@ -1,114 +1,140 @@
...
+                boundingBox: { lx: 0, ly: 1, ux: 2, uy: 3 },
+                scale: 1.0,
+                pointAttributes: ['POSITION', 'RGB'],
+                octreeDir: 'eglise_saint_blaise_arles',
+            };
+
+            const source = new PotreeSource({
+                file: 'eglise_saint_blaise_arles.json',
+                url: 'https://raw.githubusercontent.com/gmaillet/dataset/master/',
+                networkOptions: process.env.HTTPS_PROXY ? { agent: new HttpsProxyAgent(process.env.HTTPS_PROXY) } : {},
```
Comment (Desplandis) [inline test/unit/potreelayerparsing.js]:
```suggestion
```

File: test/unit/potreelayerparsing.js [hunk 12]
```diff
@@ -1,114 +1,140 @@
...
-        const layer2 = new PotreeLayer('pointsCloud2', { source, crs: view.referenceCrs });
-        layers.push(layer2);
-        const p2 = layer2.whenReady.then((l) => {
-            assert.ok(l.material.defines.NORMAL);
-            assert.ok(!l.material.defines.NORMAL_SPHEREMAPPED);
-            assert.ok(!l.material.defines.NORMAL_OCT16);
+            const source = new PotreeSource({
+                file: 'eglise_saint_blaise_arles.json',
+                url: 'https://raw.githubusercontent.com/gmaillet/dataset/master/',
+                networkOptions: process.env.HTTPS_PROXY ? { agent: new HttpsProxyAgent(process.env.HTTPS_PROXY) } : {},
```
Comment (Desplandis) [inline test/unit/potreelayerparsing.js]:
```suggestion
```

File: test/unit/3dtileslayerprocess.js [hunk 13]
```diff
@@ -1,34 +1,68 @@
...
+import sinon from 'sinon';
+import Fetcher from 'Provider/Fetcher';
 import Renderer from './bootstrap';
 
+import tilesetDiscretLOD from '../data/3dTiles/tilesetDiscretLOD.json';
+
+const urlDragonLow = 'https://raw.githubusercontent.com/CesiumGS/3d-tiles-samples/master/1.0/TilesetWithDiscreteLOD/dragon_low.b3dm';
+const urlDragonMedium = 'https://raw.githubusercontent.com/CesiumGS/3d-tiles-samples/master/1.0/TilesetWithDiscreteLOD/dragon_medium.b3dm';
+
+const dragonLow = fs.readFileSync('./test/data/3dTiles/dragon_low.b3dm');
```
Comment (Desplandis) [inline test/unit/3dtileslayerprocess.js]: Use a relative path.

```suggestion
const dragonLow = fs.readFileSync('../data/3dTiles/dragon_low.b3dm');
```

File: test/unit/3dtileslayerprocess.js [hunk 14]
```diff
@@ -1,34 +1,68 @@
...
+const dragonMedium = fs.readFileSync('./test/data/3dTiles/dragon_medium.b3dm');
```
Comment (Desplandis) [inline test/unit/3dtileslayerprocess.js]: Use a relative path.

```suggestion
const dragonMedium = fs.readFileSync('../data/3dTiles/dragon_medium.b3dm');
```

File: test/data/entwine/ept.json [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+        { "name": "Green", "type": "unsigned", "size": 2 },
+        { "name": "Blue", "type": "unsigned", "size": 2 },
+        { "name": "OriginId", "type": "unsigned", "size": 4 }
+    ],
+    "span" : 256,
+    "srs": {
+        "authority": "EPSG",
+        "horizontal": "3857",
+        "vertical": "5703",
+        "wkt": "PROJCS[\"WGS 84 ... AUTHORITY[\"EPSG\",\"3857\"]]"
```
Comment (Desplandis) [inline test/data/entwine/ept.json]: This is not a valid WKT, but we could not be testing its validity.

File: test/unit/3dtileslayerprocess.js [hunk 16]
```diff
@@ -1,34 +1,68 @@
...
 import { HttpsProxyAgent } from 'https-proxy-agent';
 import Coordinates from 'Core/Geographic/Coordinates';
+import sinon from 'sinon';
+import Fetcher from 'Provider/Fetcher';
 import Renderer from './bootstrap';
 
+import tilesetDiscretLOD from '../data/3dTiles/tilesetDiscretLOD.json';
+
+const urlDragonLow = 'https://raw.githubusercontent.com/CesiumGS/3d-tiles-samples/master/1.0/TilesetWithDiscreteLOD/dragon_low.b3dm';
+const urlDragonMedium = 'https://raw.githubusercontent.com/CesiumGS/3d-tiles-samples/master/1.0/TilesetWithDiscreteLOD/dragon_medium.b3dm';
```
Comment (Desplandis) [inline test/unit/3dtileslayerprocess.js]: It only test if the subdivision is pending, we could give it a random b3dm (or `Box.gltf`)? This would eliminate 400kB of binary files (`dragon_low.b3dm` and `dragon_medium.b3dm`).

File: test/unit/entwine.js [hunk 17]
```diff
@@ -6,12 +6,42 @@ import Coordinates from 'Core/Geographic/Coordinates';
...
-        networkOptions: process.env.HTTPS_PROXY ? { agent: new HttpsProxyAgent(process.env.HTTPS_PROXY) } : {},
+    let source;
+    let stubFetcherJson;
+    let stubFetcherArrayBuf;
+
+    before(function () {
+        stubFetcherJson = sinon.stub(Fetcher, 'json')
+            .callsFake(url => Promise.resolve(JSON.parse(resources[url])));
+        stubFetcherArrayBuf = sinon.stub(Fetcher, 'arrayBuffer')
+            .callsFake(() => Promise.resolve({}));
```
Comment (Desplandis) [inline test/unit/entwine.js]:
```suggestion
            .callsFake(() => Promise.resolve(new ArrayBuffer()));
```

File: test/unit/source.js [hunk 18]
```diff
@@ -12,11 +12,13 @@ import C3DTilesSource from 'Source/C3DTilesSource';
 import C3DTilesIonSource from 'Source/C3DTilesIonSource';
 import Extent from 'Core/Geographic/Extent';
 import { HttpsProxyAgent } from 'https-proxy-agent';
+import sinon from 'sinon';
+import Fetcher from 'Provider/Fetcher';
 
-describe('Sources', function () {
-    // geojson url to parse
-    const urlGeojson = 'https://raw.githubusercontent.com/gregoiredavid/france-geojson/master/departements/09-ariege/departement-09-ariege.geojson';
+import tileset from '../data/3dTiles/tileset.json';
```
Comment (Desplandis) [inline test/unit/source.js]:
```suggestion
const tileset = {};
```

File: test/unit/lasparser.js [hunk 19]
```diff
@@ -1,22 +1,19 @@
...
-import { HttpsProxyAgent } from 'https-proxy-agent';
+import fs from 'fs';
 import LASParser from 'Parser/LASParser';
-import Fetcher from 'Provider/Fetcher';
 import { compareWithEpsilon } from './utils';
 
+const urlData = './test/data/pointclouds/';
+const lasFile = fs.readFileSync(`${urlData}/data_test.las`);
+const lazFile = fs.readFileSync(`${urlData}/data_test.laz`);
+const lazV14File = fs.readFileSync(`${urlData}/ellipsoid-1.4.laz`);
```
Comment (Desplandis) [inline test/unit/lasparser.js]: Those two binary files have a total size of 619kB, I'll try to generate another dataset!

**Review by Desplandis — COMMENTED** [review]
List of suggestions in this PR. I'll try to generate smaller datasets for potree and laz (1.2 & 1.4) files!

File: test/unit/3dtileslayerprocessbatchtable.js (hunk 1 above)
Comment (ftoromanoff) [reply test/unit/3dtileslayerprocessbatchtable.js]: I tried to do so but (at least on my computer) it's not working. It might be linked with my platform.

Comment (ftoromanoff) [reply test/unit/3dtileslayerprocessbatchtable.js]: Done

File: test/unit/3dtileslayerstyle.js (hunk 2 above)
Comment (ftoromanoff) [reply test/unit/3dtileslayerstyle.js:38]: I did not write this part of the code just moved it from the it() to the main scope.

File: test/unit/3dtileslayerstyle.js (hunk 3 above)
Comment (ftoromanoff) [reply test/unit/3dtileslayerstyle.js]: Removed

File: test/unit/3dtileslayerstyle.js (hunk 4 above)
Comment (ftoromanoff) [reply test/unit/3dtileslayerstyle.js]: Done

File: test/unit/3dtileslayerstyle.js (hunk 5 above)
Comment (ftoromanoff) [reply test/unit/3dtileslayerstyle.js]: Done. And tileset.json file removed.

File: test/unit/geoidlayer.js (hunk 7 above)
Comment (ftoromanoff) [reply test/unit/geoidlayer.js]: see next remark.

File: test/unit/gtxparser.js (hunk 8 above)
Comment (ftoromanoff) [reply test/unit/gtxparser.js]: I move it to bootstrap.js

File: test/unit/potree.js (hunk 10 above)
Comment (ftoromanoff) [reply test/unit/potree.js]: suppresed

File: test/unit/potreelayerparsing.js (hunk 11 above)
Comment (ftoromanoff) [reply test/unit/potreelayerparsing.js]: same.

File: test/unit/potreelayerparsing.js (hunk 12 above)
Comment (ftoromanoff) [reply test/unit/potreelayerparsing.js]: same

File: test/unit/3dtileslayerprocess.js (hunk 13 above)
Comment (ftoromanoff) [reply test/unit/3dtileslayerprocess.js]: already answered

File: test/unit/3dtileslayerprocess.js (hunk 14 above)
Comment (ftoromanoff) [reply test/unit/3dtileslayerprocess.js]: same

File: test/data/entwine/ept.json (hunk 15 above)
Comment (ftoromanoff) [reply test/data/entwine/ept.json]: I did the change in a futur PR #2272 as i will need it, but indeed I can do now.

File: test/unit/entwine.js (hunk 17 above)
Comment (ftoromanoff) [reply test/unit/entwine.js]: Indeed, even if we never currently test the returned data, It will be better to have a valid return.

File: test/unit/source.js (hunk 18 above)
Comment (ftoromanoff) [reply test/unit/source.js]: changed (see a previous remark)

File: src/Layer/C3DTilesLayer.js [hunk 20]
```diff
@@ -30,8 +30,8 @@ const update = process3dTilesNode();
 
 /**
  * Find tileId of object
- *
```
Comment (AnthonyGlt) [inline src/Layer/C3DTilesLayer.js:33]: why this removal of * everytime ? Is it the way to format this kind of comments ?

File: test/unit/3dtileslayerprocessbatchtable.js [hunk 21]
```diff
@@ -1,49 +1,85 @@
...
+        let stubFetcherJson;
+        let viewer;
+        let threedTilesLayerBTHierarchy;
+
+        before(function () {
+            stubFetcherJson = sinon.stub(Fetcher, 'json')
+                .callsFake(() => Promise.resolve(tileset));
+
+            stubFetcherArrayBuf = sinon.stub(Fetcher, 'arrayBuffer')
+                .callsFake(() => Promise.resolve(b3dm));
```
Comment (AnthonyGlt) [inline test/unit/3dtileslayerprocessbatchtable.js:41]: Why do you use `stub` here when we had to fetch the resource above ? l25
Is it, in case of failure, to skip directly the test ?

File: test/unit/dataSourceProvider.js [hunk 22]
```diff
@@ -103,6 +104,8 @@ describe('Provide in Sources', function () {
     context.elevationLayers = [elevationlayer];
     context.colorLayers = [colorlayer];
 
+    stub.restore();
```
Comment (AnthonyGlt) [inline test/unit/dataSourceProvider.js]: why restore the other tests ? Should be after no ?
If not, then why mock `['image/png', () => Promise.resolve(new THREE.Texture())],`, it's not use right now ?

File: test/unit/demutils.js [hunk 23]
```diff
@@ -12,39 +12,66 @@ import OBB from 'Renderer/OBB';
...
-    const elevationlayer = new ElevationLayer('worldelevation', { source });
+    let elevationlayer;
+    let context;
+    let stubSuppFetcher;
+    const ELEVATION = 300;
+
+    before(function () {
+        stubSuppFetcher = sinon.stub(supportedFetchers, 'get')
+            .callsFake((format) => {
+                if (format === 'image/x-bil;bits=32') {
```
Comment (AnthonyGlt) [inline test/unit/demutils.js]:
```suggestion
        stubSuppFetcher = sinon.stub(supportedFetchers, 'get');
        stubSuppFetcher.withArgs('image/x-bil;bits=32')
            .callsFake(() => {
                const floatArray = createBilData(ELEVATION);
```
Or is it meant to throw an error ? This way the mock happen only for this format. WDYT ?

File: test/unit/featuregeometrylayer.js [hunk 24]
```diff
@@ -3,71 +3,85 @@ import assert from 'assert';
...
-    const source2 = new FileSource({
-        url: 'https://raw.githubusercontent.com/gregoiredavid/france-geojson/master/departements/09-ariege/departement-09-ariege.geojson',
-        crs: 'EPSG:4326',
-        format: 'application/json',
-        networkOptions: process.env.HTTPS_PROXY ? { agent: new HttpsProxyAgent(process.env.HTTPS_PROXY) } : {},
-    });
+    before(function () {
+        stubSuppFetcher = sinon.stub(supportedFetchers, 'get')
+            .callsFake((format) => {
+                if (format === 'application/json') {
```
Comment (AnthonyGlt) [inline test/unit/featuregeometrylayer.js]: same question, limit call using `withArgs` ? I don't know which is best

File: test/unit/entwine.js [hunk 25]
```diff
@@ -1,17 +1,45 @@
...
 import EntwinePointTileNode from 'Core/EntwinePointTileNode';
+import sinon from 'sinon';
+import Fetcher from 'Provider/Fetcher';
 import Renderer from './bootstrap';
 
+import ept from '../data/entwine/ept.json';
+import eptHierarchy from '../data/entwine/ept-hierarchy/0-0-0-0.json';
+
+const urlEpt = 'https://raw.githubusercontent.com/iTowns/iTowns2-sample-data/master/pointclouds/entwine/ept.json';
+const urlEptHierarchy = 'https://raw.githubusercontent.com/iTowns/iTowns2-sample-data/master/pointclouds/entwine/ept-hierarchy/0-0-0-0.json';
```
Comment (AnthonyGlt) [inline test/unit/entwine.js]: Maybe declare another const for `https://raw.githubusercontent.com/iTowns/iTowns2-sample-data/master/pointclouds/entwine/`

File: test/unit/orientedimagelayer.js [hunk 26]
```diff
@@ -2,69 +2,94 @@ import proj4 from 'proj4';
...
 import GlobeView from 'Core/Prefab/GlobeView';
+import sinon from 'sinon';
+import Fetcher from 'Provider/Fetcher';
 import Renderer from './bootstrap';
 
+import panoData from '../data/OrientedImage/panoramicsMetaDataParis.geojson';
+import camCalibration from '../data/OrientedImage/cameraCalibration.json';
+
+const orientationsUrl = 'https://raw.githubusercontent.com/iTowns/iTowns2-sample-data/master/immersive/exampleParis1/panoramicsMetaDataParis.geojson';
+const calibrationUrl = 'https://raw.githubusercontent.com/iTowns/iTowns2-sample-data/master/immersive/exampleParis1/cameraCalibration.json';
```
Comment (AnthonyGlt) [inline test/unit/orientedimagelayer.js]: Maybe declare an other const for `https://raw.githubusercontent.com/iTowns/iTowns2-sample-data/master/immersive/exampleParis1/`

This comment applied to multiple files

File: src/Layer/C3DTilesLayer.js (hunk 20 above)
Comment (ftoromanoff) [reply src/Layer/C3DTilesLayer.js:33]: From my point of view, yes.
There is :
```js
* Description
* @param1
* @param2
* #emptyline
* @returns
```
Mostly to have a delimiter between param and returns for lisibility

File: test/unit/3dtileslayerprocessbatchtable.js (hunk 21 above)
Comment (ftoromanoff) [reply test/unit/3dtileslayerprocessbatchtable.js:41]: yes, It's to separate the fetching that can fail for various reasons, independently to the features that we want to test and the tests. In the of case the fetching had failed, we skip all the tests and only the message: Fetching failed remains.

File: src/Layer/C3DTilesLayer.js (hunk 20 above)
Comment (AnthonyGlt) [reply src/Layer/C3DTilesLayer.js:33]: but you don't keep the asterisk ?

File: test/unit/source.js [hunk 27]
```diff
@@ -283,38 +306,39 @@ describe('Sources', function () {
...
     });
 
     describe('C3DTilesSource', function () {
-        const params3DTiles = {
-            url: 'https://raw.githubusercontent.com/iTowns/iTowns2-sample-data/master/3DTiles/lyon_1_4978/tileset.json',
-            networkOptions: process.env.HTTPS_PROXY ? { agent: new HttpsProxyAgent(process.env.HTTPS_PROXY) } : {},
-        };
+        let stubFetcherJson;
+        before(function () {
+            stubFetcherJson = sinon.stub(Fetcher, 'json')
```
Comment (jailln) [inline test/unit/source.js:316]: I think we could factorize a bit this mocking logic in one place (e.g. in `utils.js`) by creating the following method:

```
function getFetcherStub(method, returnVal) {
    sinon.stub(Fetcher, method)
    .callsFake(() => Promise.resolve(returnVal));
}
``` 
And just do `stubFetcherJson = utils.getFetcherStub('json', tileset)` here (and use this method everywhere else where you create stubs). WDYT?

File: src/Layer/C3DTilesLayer.js (hunk 20 above)
Comment (ftoromanoff) [reply src/Layer/C3DTilesLayer.js:33]: I removed the empty line between the description and the list of params.

File: test/unit/dataSourceProvider.js (hunk 22 above)
Comment (ftoromanoff) [reply test/unit/dataSourceProvider.js]: I move everything in a before. We need to mock the fetching of json and image as we insteancate many layers: PlanarLayer, colorLayer, ElevationLayer and GeometryLayer.
But later on we will need to mock these (te creation of Layer) instead of mocking the fetching used in the layer creation.

File: test/unit/demutils.js (hunk 23 above)
Comment (ftoromanoff) [reply test/unit/demutils.js]: I didn't know that this method was existing. I did the change.
If the tests are well done, there shouldn't be a nned for throwing an error.

File: test/unit/featuregeometrylayer.js (hunk 24 above)
Comment (ftoromanoff) [reply test/unit/featuregeometrylayer.js]: done (and in all other test files)

File: test/unit/entwine.js (hunk 25 above)
Comment (ftoromanoff) [reply test/unit/entwine.js]: Done

File: test/unit/orientedimagelayer.js (hunk 26 above)
Comment (ftoromanoff) [reply test/unit/orientedimagelayer.js]: Done. In several files.

File: test/unit/source.js (hunk 27 above)
Comment (ftoromanoff) [reply test/unit/source.js:316]: The purpose of this PR is to get in the direction of real unti tests. Thus i did mock the fetching everywhere, but later on we should be mocking the process that use the mocked fetcher (except for some specifique tests).That's why I think that a it will be on some further PR to think about how to factorize the different mocking parts.

Comment (jailln) [reply test/unit/source.js:316]: Ok, in this case could you open an issue about it please ?

File: test/unit/orientedimagelayer.js [hunk 28]
```diff
@@ -2,69 +2,97 @@ import proj4 from 'proj4';
...
-        { renderer,
-            noControls: true,
-            handleCollision: false,
-            sseSubdivisionThreshold: 10,
+        // Prepare oriented image source
+        const orientedImageSource = new OrientedImageSource({
+            url: 'http://www.itowns-project.org/itowns-sample-data-small/images/140616/Paris-140616_0740-{cameraId}-00001_0000{panoId}.jpg',
+            // Url to a GEOJSON file describing feature points. It describes position and orientation of each panoramic.
+            orientationsUrl: 'https://raw.githubusercontent.com/iTowns/iTowns2-sample-data/master/' +
+                'immersive/exampleParis1/panoramicsMetaDataParis.geojson',
```
Comment (AnthonyGlt) [inline test/unit/orientedimagelayer.js:50]: You could use `baseurl` const

File: test/unit/fetcher.js [hunk 29]
```diff
@@ -0,0 +1,165 @@
+import assert from 'assert';
+import { HttpsProxyAgent } from 'https-proxy-agent';
+import Fetcher from 'Provider/Fetcher';
+import { DataTexture, Texture } from 'three';
+
+describe('Fetcher', function () {
+    const networkOptions = process.env.HTTPS_PROXY ? { agent: new HttpsProxyAgent(process.env.HTTPS_PROXY) } : {};
+    describe('text', function () {
+        const url = 'https://raw.githubusercontent.com/iTowns/iTowns2-sample-data/master/' +
+        'altitude-conversion-grids/EGM2008.gdf';
```
Comment (AnthonyGlt) [inline test/unit/fetcher.js]: Could define a `baseUrl` like you did elsewhere, to use here, and in l.24, l.45 etc..

File: test/unit/orientedimagelayer.js (hunk 28 above)
Comment (ftoromanoff) [reply test/unit/orientedimagelayer.js:50]: In my opinon, this url is a parameter of ```new OrientedImageSource()``` thus it's a string and will not really be interpreted as an url (url: 'urlOrientedImage.jpg' will have work the same way). I don't really see the use of trying to reduce it at a base url name.

File: test/unit/fetcher.js (hunk 29 above)
Comment (ftoromanoff) [reply test/unit/fetcher.js]: Done

**Review by AnthonyGlt — APPROVED** [review]
